### PR TITLE
Reworked FloydWarshallShortestPaths code, updated As(Un)Weighted(Directed)Graph classes, added a Mixed-Graph Implementation.

### PR DESCRIPTION
**Pull request by jkinable**
Revision of the FloydWarshallShortestPaths implementation. I made several performance improvements while simultaneously reducing the amount of code.

Improvements:
-Replaced recursive algorithm by simple, more efficient loop.
-For some reason, if the user would query the shortest path from i to j, the previous version of the code would first calculate ALL shortest paths for every possible i,j pair and then return the requested one. This is highly inefficient if the user only needs a few (which is almost always the case). 
-Removed a lot of unnecessary code

Made some changes to the As(Un)Weighted(Directed)Graph classes. The AsUnweightedDirectedGraph now simply extends the AsUnweightedGraph class, thereby reducing code. Furthermore, I've added the AsWeightedDirectedGraph which did not exist (the AsWeightedGraph class already existed).

Added a Mixed-Graph implementation to support graphs which have both edges and arcs. Furthermore, replaced the HashSets and HashMaps in GraphUnion.java with their Linked equivalents.

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/FloydWarshallShortestPaths.java [hunk 1]
```diff
@@ -128,18 +133,20 @@ private void lazyCalculateMatrix()
...
+            for (E edge : graph.edgeSet()) {
+                int v_1 = vertexIndices.get(graph.getEdgeSource(edge));
+                int v_2 = vertexIndices.get(graph.getEdgeTarget(edge));
+                d[v_1][v_2] =d[v_2][v_1] =graph.getEdgeWeight(edge);
+            }
+        }else{ //This works for both Directed and Mixed graphs! Iterating over the arcs and querying source/sink does not suffice for graphs which contain both edges and arcs
+            DirectedGraph<V,E> directedGraph=(DirectedGraph<V,E>)graph;
+            for(V v1 : directedGraph.vertexSet()){
+                for(V v2 : Graphs.successorListOf(directedGraph, v1)){
+                    int v_1 = vertexIndices.get(v1);
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/FloydWarshallShortestPaths.java]: Pull this out one level.

File: jgrapht-core/src/main/java/org/jgrapht/graph/AsWeightedDirectedGraph.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package org.jgrapht.graph;
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/graph/AsWeightedDirectedGraph.java]: All files need the standard JGraphT header.

**Comment by jsichi** [comment]
This pull requests needs tests for the new classes.

**Comment by jkinable** [comment]
Hi John, I've made the requested changes and I've added test classes for the AsWeightedDirectedGraph, UndirectedGraphUnion, DirectedGraphUnion, MixedGraphUnion. Notice that the  UndirectedGraphUnion and DirectedGraphUnion classes were already present in previous version of JgraphT, but no test classes were included.

**Comment by jkinable** [comment]
Unless no further requests, I think this class can be merged.

**Comment by jsichi** [comment]
Thanks for leaving everything better than you found it!

**Comment by jsichi** [comment]
(In a followup commit, I've added @Deprecated on the erroneous UndirectedGraphUnion constructors (I really doubt anyone is using them!)

**Pull request merged**